[clang-tidy] False positives bugprone-dynamic-static-initializers

I believe the check warns incorrectly in the following cases:

```cpp
#include <limits>

template <typename T>
constexpr T kMin{std::numeric_limits<T>::min()};

template <int n>
struct X
{
  static constexpr int xn{n};
};
```

```bash
$ clang-tidy --checks=-*,bugprone-dynamic-static-initializers foo.hpp -- -fno-threadsafe-statics
589 warnings generated.
/tmp/foo.hpp:4:13: warning: static variable 'kMin' may be dynamically initialized in this header file [bugprone-dynamic-static-initializers]
    4 | constexpr T kMin{std::numeric_limits<T>::min()};
      |             ^
/tmp/foo.hpp:9:24: warning: static variable 'xn' may be dynamically initialized in this header file [bugprone-dynamic-static-initializers]
    9 |   static constexpr int xn{n};
      |                        ^

```

--- a/clang-tools-extra/clang-tidy/bugprone/DynamicStaticInitializersCheck.cpp
+++ b/clang-tools-extra/clang-tidy/bugprone/DynamicStaticInitializersCheck.cpp
@@ -18,12 +18,11 @@ namespace clang::tidy::bugprone {
 namespace {
 
 AST_MATCHER(clang::VarDecl, hasConstantDeclaration) {
+  if (Node.isConstexpr())
+    return true;
   const Expr *Init = Node.getInit();
-  if (Init && !Init->isValueDependent()) {
-    if (Node.isConstexpr())
-      return true;
+  if (Init && !Init->isValueDependent())
     return Node.evaluateValue();
-  }
   return false;
 }
 

--- a/clang-tools-extra/docs/ReleaseNotes.rst
+++ b/clang-tools-extra/docs/ReleaseNotes.rst
@@ -370,6 +370,11 @@ Changes in existing checks
   new option `IgnoreMacros` to suppress warnings within macro
   expansions.
 
+- Improved :doc:`bugprone-dynamic-static-initializers
+  <clang-tidy/checks/bugprone/dynamic-static-initializers>` check by
+  avoiding false positives for ``constexpr`` variables whose
+  initializers are value-dependent.
+
 - Improved :doc:`bugprone-easily-swappable-parameters
   <clang-tidy/checks/bugprone/easily-swappable-parameters>` check by
   correcting a spelling mistake on its option

--- a/clang-tools-extra/test/clang-tidy/checkers/bugprone/dynamic-static-initializers.hpp
+++ b/clang-tools-extra/test/clang-tidy/checkers/bugprone/dynamic-static-initializers.hpp
@@ -42,3 +42,24 @@ int foo3() {
   static int p = 7 + 83; // no warning
   return p;
 }
+
+namespace std {
+  template <typename T>
+  struct numeric_limits {
+    static constexpr T min() { return 0; }
+    static constexpr T max() { return 1000; }
+  };
+}
+
+template <typename T>
+void template_func() {
+  static constexpr T local_kMin{std::numeric_limits<T>::min()}; // no warning
+}
+
+template <int n>
+struct TemplateStruct {
+  static constexpr int xn{n}; // no warning
+};
+
+template <typename T>
+constexpr T kGlobalMin{std::numeric_limits<T>::min()}; // no warning

